src/Routing.jl

```diff
@@ -1,66 +1,76 @@
-export compute_streamflow!,
-    update_state_from_hillslope!, update_state_from_channel!
+export compute_streamflow,
+    compute_river_state, compute_channel_state, compute_hillslope_state
 
 using CSV, DataFrames, Dates, DSP, SpecialFunctions
 
 # methods
-function compute_streamflow!(
-    river_state::RS,
+function compute_streamflow(
+    initial_date_window::DateWindow,
     river_model::HCM,
     env::E,
-    start_date::Date,
     end_date::Date,
-) where {RS <: RiverState, HCM <: HillslopeChannelRiverModel, E <: Environment}
-    update_state!(river_state, river_model, env, start_date, end_date)
-    return compute_streamflow(river_state, env)
+) where {HCM <: HillslopeChannelRiverModel, E <: Environment}
+
+    date_window = initial_date_window
+    river_states = []
+    streamflows = []
+    while date_window.end_date <= end_date
+        # @info "computing streamflow over window [$(date_window.start_date),$(date_window.end_date)]"
+        river_state = compute_river_state(date_window, river_model, env)
+        push!(river_states, river_state)
+        push!(streamflows, compute_streamflow(river_state, env))
+        date_window = iterate(date_window)
+    end
+
+    return streamflows, river_states
 end
 
-function update_state!(
-    river_state::RS,
+function compute_river_state(
+    date_window::DateWindow,
     river_model::HCM,
     env::E,
-    start_date::Date,
-    end_date::Date,
-) where {RS <: RiverState, HCM <: HillslopeChannelRiverModel, E <: Environment}
-    update_state_from_hillslope!(
-        river_state,
-        river_model.hillslope_model,
-        env,
-        start_date,
-        end_date,
-    )
-    update_state_from_channel!(
-        river_state,
+) where {HCM <: HillslopeChannelRiverModel, E <: Environment}
+
+    new_hillslope_state =
+        compute_hillslope_state(date_window, river_model.hillslope_model, env)
+
+    new_channel_state = compute_channel_state(
+        new_hillslope_state,
+        date_window,
         river_model.channel_model,
         env,
-        start_date,
-        end_date,
     )
+
+    return HillslopeChannelRiverState(
+        new_hillslope_state,
+        new_channel_state,
+        date_window,
+    )
+
+
 end
 
-function update_state_from_hillslope!(
-    river_state::RS,
+function compute_hillslope_state(
+    date_window::DateWindow,
     hillslope_model::HM,
     static_env::SE,
     dynamic_env::DE,
-    start_date::Date,
-    end_date::Date,
 ) where {
-    RS <: RiverState,
     HM <: AbstractHillslopeModel,
     SE <: StaticEnvironment,
     DE <: DynamicEnvironment,
 }
-    println("Starting hillslope update")
-
     # Constants
     day_to_s = 86400
     km²_to_m² = 1000000
 
     forcing_timeseries = dynamic_env.forcing_timeseries
     output_dir = dynamic_env.output_dir
-    dates = collect(start_date:Day(1):end_date)
+
+    start_date = date_window.start_date
+    end_date = date_window.end_date
+    dates = get_all_dates(date_window)
```

Comment:
I think this should just be the dates from current t to current t - t_max_hillslope. Otherwise we read in all of the forcing data (OK for now), but then take the convolution over all forcing times, right? But we really only need the convolution over the past t_max_hillslope times.

Reply:
So this date-window is supposed to represent just the dynamic environment, dates and the frequency of data. It is decoupled from the computation window. i.e. this is just a way of recording all the dates in the forcing timeseries and is not used in the code, except to provide the start and end time of the "while loop" of the simulation. 

I think it is good to keep the window we hold data decoupled from the window we use to compute river state update.

Reply:
I think maybe we do need this for all times. Because we don't know how far back in time the downstream needs to compute right? Like if you only do it at `t-t_hillslope`, but then you need to compute q(t) at some downstream basin which is another 5 days away, then wouldn't we need information at `t-5-t_hillslope` to compute that? (because we need the hillslope contribution at `t-5` at the upstream basin which has a `t-hillslope`

Maybe there is a more hierarchical way we could handle the date-windows to help here rather than what I do now, which is just a long history window.